packages/vite/src/node/optimizer/scan.ts

```diff
@@ -106,19 +106,33 @@ export async function scanImports(config: ResolvedConfig): Promise<{
   const { plugins = [], ...esbuildOptions } =
     config.optimizeDeps?.esbuildOptions ?? {}
 
-  await build({
-    absWorkingDir: process.cwd(),
-    write: false,
-    stdin: {
-      contents: entries.map((e) => `import ${JSON.stringify(e)}`).join('\n'),
-      loader: 'js',
-    },
-    bundle: true,
-    format: 'esm',
-    logLevel: 'error',
-    plugins: [...plugins, plugin],
-    ...esbuildOptions,
-  })
+  try {
+    await build({
+      absWorkingDir: process.cwd(),
+      write: false,
+      stdin: {
+        contents: entries.map((e) => `import ${JSON.stringify(e)}`).join('\n'),
+        loader: 'js',
+      },
+      bundle: true,
+      format: 'esm',
+      logLevel: 'silent',
+      plugins: [...plugins, plugin],
+      ...esbuildOptions,
+    })
+  } catch (e) {
+    const err = new Error('Failed to scan dependencies from entries')
+    err.stack = e.stack + '\n' + err.stack
+    const esbuildMsg = e.errors
+      ? await formatMessages(e.errors, { kind: 'error' })
+      : ''
+
+    config.logger.error(
+      colors.red(
+        `${err.message}:\n${entries.join('\n')}\n\n${esbuildMsg}${err.stack}`,
+      ),
+    )
```

Comment:
IIUC this re-logs out the error instead of throwing, I think we may still want to throw to stop the execution? I think we can follow a similar pattern like:

https://github.com/vitejs/vite/blob/a613d45d2684e51a09f64a19b36480f1f301788b/packages/vite/src/node/plugins/css.ts#L1363-L1369

Not the `e.errors` part, but the pattern where it mutates `e` to inject extra information of the error, in our case, mutating `.message` with the `Failed to scan dependencies from entries` part.

Reply:
I prefer to throw, too. Logging error instead of throw is the previous practice so I didn't change it here. Currently, all the throw is covered up by

https://github.com/vitejs/vite/blob/a613d45d2684e51a09f64a19b36480f1f301788b/packages/vite/src/node/optimizer/optimizer.ts#L237-L240

So we need to change the catch block if we want to throw to terminate the process which means we need to distinguish which errors should terminate the process and which are not. For this PR, we may need to distinguish which errors are thrown when scanning the HTML. Am I understanding this correctly? 🤔

Reply:
Do you mean the catch above the finally block at 

https://github.com/vitejs/vite/blob/a613d45d2684e51a09f64a19b36480f1f301788b/packages/vite/src/node/optimizer/optimizer.ts#L235-L240

I don't think we need to terminate the process if scanning failed like before. We only need the throw to skip the part below it

https://github.com/vitejs/vite/blob/a873af5d3d0b59ee8c1f982ad6ff821b7119c8fb/packages/vite/src/node/optimizer/scan.ts#L123-L129

so that the function doesn't appear to be successful when it errors. Currently since this PR doesn't throw, the linked code would still run, which shouldn't.

Reply:
Mutated the origin e from the catch block and throw it while still using the `logger.error` to log the error message from esbuild and created a `new Error` to compose the complete error stack as shown in the image of the PR description.